### ISSUE TITLE
Binaryen API discarding import function names when reading

### ISSUE THREAD

**Issue by AmaanC** [WebAssembly/binaryen#1285]
`name-mangling.wast`:

```
(module
  (type (;0;) (func (param i32 i32)))
  (type (;1;) (func))
  (type (;2;) (func (param i32) (result i32)))
  (type (;3;) (func (result i32)))
  (import "env" "memoryBase" (global (;0;) i32))
  ;; What really matters here:
  (import "env" "_my_external_func" (func $_my_external_func (type 1)))
  (import "env" "memory" (memory (;0;) 256))
  (import "env" "table" (table (;0;) 0 anyfunc))
  (import "env" "tableBase" (global (;1;) i32))
)
```

`name-mangling.html` ([live](http://whatthedude.com/binaryen-demo/name-mangling.html)):

```
<script src='https://rawgit.com/WebAssembly/binaryen/master/bin/binaryen.js'></script>
<script>
    function download(file_or_blob, name)
    {
        var a = document.createElement("a");
        a["download"] = name;
        a.href = window.URL.createObjectURL(file_or_blob);
        a.dataset["downloadurl"] = ["application/octet-stream", a["download"], a.href].join(":");

        if(document.createEvent)
        {
            var ev = document.createEvent("MouseEvent");
            ev.initMouseEvent("click", true, true, window,
                0, 0, 0, 0, 0, false, false, false, false, 0, null);
            a.dispatchEvent(ev);
        }
        else
        {
            a.click();
        }

        window.URL.revokeObjectURL(a.href);
    }
    function dump_file(ab, name)
    {
        if(!(ab instanceof Array))
        {
            ab = [ab];
        }

        var blob = new Blob(ab);
        download(blob, name);
    }

    fetch('name-mangling.wasm').then((resp) => {
        return resp.arrayBuffer();
    }).then((buffer) => {
        var module = Binaryen.readBinary(new Uint8Array(buffer));
        console.log(module.validate(module.emitBinary()));
        dump_file(module.emitBinary(), 'name-mangled.wasm');
    });
</script>
```

Running `wasm2wast` on `name-mangled.wasm` after download:

```
(module
  (type (;0;) (func (param i32 i32)))
  (type (;1;) (func))
  (type (;2;) (func (param i32) (result i32)))
  (type (;3;) (func (result i32)))
  (import "env" "memoryBase" (global (;0;) i32))
  ;; Note how the internalName for the import is dropped and changed to `$import$1`
  (import "env" "_my_external_func" (func $import$1 (type 1)))
  (import "env" "memory" (memory (;0;) 256))
  (import "env" "table" (table (;0;) 0 anyfunc))
  (import "env" "tableBase" (global (;1;) i32)))
```

Note how `$_my_external_func` was changed to `$import$1` here.

The fact that this happens makes it harder to use binaryen.js to remove `import`s from existing wasm binaries since it makes the internalName for the import unpredictable.

Likely cause in the code:
https://github.com/WebAssembly/binaryen/blob/master/src/wasm/wasm-binary.cpp#L1520

**Comment by kripken** [comment]
I believe wasm doesn't have a way to provide names for imports in the binary format, see "Name Section" here: http://webassembly.org/docs/binary-encoding/ - it only supports names for modules, functions, and locals, but not imports.

**Comment by binji** [comment]
You can name function imports, see http://webassembly.github.io/spec/appendix/custom.html#function-names. It maps names to function indices, which includes the function imports first.

**Comment by kripken** [comment]
Oh, thanks. Then this is indeed a bug we should fix.

**Comment by kripken** [comment]
Fix in #1290

**Comment by kripken** [comment]
Fixed by that PR.
